Investigate and fix semaphore test timeout

See [failure here](https://github.com/letsencrypt/boulder/actions/runs/3887811021/jobs/6634842506#step:6:374):

```
panic: test timed out after 10m0s

goroutine 13 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:2036 +0xbb
created by time.goFunc
	/usr/local/go/src/time/sleep.go:176 +0x48

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000009860, {0x5d0345, 0xe}, 0x5d9db8)
	/usr/local/go/src/testing/testing.go:1494 +0x789
testing.runTests.func1(0x0?)
	/usr/local/go/src/testing/testing.go:[184](https://github.com/letsencrypt/boulder/actions/runs/3887811021/jobs/6634842506#step:6:185)6 +0x9a
testing.tRunner(0xc000009860, 0xc000071ba0)
	/usr/local/go/src/testing/testing.go:1446 +0x217
testing.runTests(0xc000002140?, {0x6e8e20, 0x8, 0x8}, {0x40?, 0x7f5898c912b0?, 0x6ed360?})
	/usr/local/go/src/testing/testing.go:1844 +0x7ed
testing.(*M).Run(0xc000002140)
	/usr/local/go/src/testing/testing.go:1726 +0xa85
main.main()
	_testmain.go:69 +0x2ea

goroutine 4 [chan receive]:
testing.(*T).Parallel(0xc000009a00)
	/usr/local/go/src/testing/testing.go:1271 +0x4f3
github.com/letsencrypt/boulder/semaphore_test.TestWeighted(0x0?)
	/boulder/semaphore/semaphore_test.go:30 +0x31
testing.tRunner(0xc000009a00, 0x5d9de0)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 5 [chan receive]:
testing.(*T).Parallel(0xc000009d40)
	/usr/local/go/src/testing/testing.go:1271 +0x4f3
github.com/letsencrypt/boulder/semaphore_test.TestWeightedPanic(0xc000009d40)
	/boulder/semaphore/semaphore_test.go:48 +0x37
testing.tRunner(0xc000009d40, 0x5d9dd0)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 6 [chan receive]:
testing.(*T).Parallel(0xc00011e1a0)
	/usr/local/go/src/testing/testing.go:1271 +0x4f3
github.com/letsencrypt/boulder/semaphore_test.TestWeightedTryAcquire(0xc00011e1a0)
	/boulder/semaphore/semaphore_test.go:60 +0x48
testing.tRunner(0xc00011e1a0, 0x5d9dd8)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 7 [chan receive]:
testing.(*T).Parallel(0xc00011e4e0)
	/usr/local/go/src/testing/testing.go:1271 +0x4f3
github.com/letsencrypt/boulder/semaphore_test.TestWeightedAcquire(0xc00011e4e0)
	/boulder/semaphore/semaphore_test.go:84 +0x48
testing.tRunner(0xc00011e4e0, 0x5d9dc0)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 8 [chan receive]:
testing.(*T).Parallel(0xc00011e820)
	/usr/local/go/src/testing/testing.go:1271 +0x4f3
github.com/letsencrypt/boulder/semaphore_test.TestWeightedDoesntBlockIfTooBig(0xc00011e820)
	/boulder/semaphore/semaphore_test.go:114 +0x48
testing.tRunner(0xc00011e820, 0x5d9dc8)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 9 [chan receive]:
testing.(*T).Parallel(0xc00011eb60)
	/usr/local/go/src/testing/testing.go:1271 +0x4f3
github.com/letsencrypt/boulder/semaphore_test.TestLargeAcquireDoesntStarve(0x0?)
	/boulder/semaphore/semaphore_test.go:143 +0x34
testing.tRunner(0xc00011eb60, 0x5d9db0)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 21 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters(0xc000084820)
	/boulder/semaphore/semaphore_test.go:225 +0x2e7
testing.tRunner(0xc000084820, 0x5d9db8)
	/usr/local/go/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x75e

goroutine 22 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:[212](https://github.com/letsencrypt/boulder/actions/runs/3887811021/jobs/6634842506#step:6:213) +0x174

goroutine 23 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:[214](https://github.com/letsencrypt/boulder/actions/runs/3887811021/jobs/6634842506#step:6:215) +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 12 [chan receive]:
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:[215](https://github.com/letsencrypt/boulder/actions/runs/3887811021/jobs/6634842506#step:6:216) +0x85
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 24 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 25 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 26 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 27 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 28 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 29 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174

goroutine 30 [select]:
github.com/letsencrypt/boulder/semaphore.(*Weighted).Acquire(0xc00008a050, {0x607810, 0xc000086040}, 0x1)
	/boulder/semaphore/semaphore.go:76 +0x5e5
github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters.func1()
	/boulder/semaphore/semaphore_test.go:214 +0x6e
created by github.com/letsencrypt/boulder/semaphore_test.TestMaxWaiters
	/boulder/semaphore/semaphore_test.go:212 +0x174
FAIL	github.com/letsencrypt/boulder/semaphore	600.105s
```

First appeared after merging https://github.com/letsencrypt/boulder/pull/6568, but did not appear in that change's pre-submit tests, and that change shouldn't affect the behavior of this test at all.

--- a/semaphore/semaphore.go
+++ b/semaphore/semaphore.go
@@ -122,6 +122,12 @@ func (s *Weighted) Release(n int64) {
 	s.mu.Unlock()
 }
 
+func (s *Weighted) NumWaiters() int {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.waiters.Len()
+}
+
 func (s *Weighted) notifyWaiters() {
 	for {
 		next := s.waiters.Front()

--- a/semaphore/semaphore_test.go
+++ b/semaphore/semaphore_test.go
@@ -207,21 +207,19 @@ func TestMaxWaiters(t *testing.T) {
 	sem := semaphore.NewWeighted(1, 10)
 	sem.Acquire(ctx, 1)
 
-	ch := make(chan struct{})
 	for i := 0; i < 10; i++ {
 		go func() {
-			ch <- struct{}{}
 			sem.Acquire(ctx, 1)
 			<-ctx.Done()
 		}()
-		// We want to wait until the goroutine is actually blocked in `sem.Acquire`
-		// but we have no way of doing that. Instead, wait for the goroutine to
-		// have written to ch. This is a race condition! But it seems to work out.
-		// If the race condition works out badly, we'll hang below on line 224
-		// when trying to acquire the semaphore (instead of erroring out early).
-		<-ch
 	}
 
+	// Since the goroutines that act as waiters are intended to block in
+	// sem.Acquire, there's no principled wait to trigger here once they're
+	// blocked. Instead, loop until we reach the expected number of waiters.
+	for sem.NumWaiters() < 10 {
+		time.Sleep(10 * time.Millisecond)
+	}
 	err := sem.Acquire(ctx, 1)
 	if err != semaphore.ErrMaxWaiters {
 		t.Errorf("expected error when maxWaiters was reached, but got %#v", err)

